game.c

```diff
@@ -15,155 +15,167 @@
 }
 
-void init_SDL_settings() {
+void init_SDL_settings() {}
+
+int init_arguments() { return 0; }
+
+int process_arguments(int ac, char **av,
+                      struct argument_parameters *arguments) {
+    return 0;
 }
 
-int init_arguments() {
-	return 0;
+long ONE_NANOSECOND = 1000000000;
+
+long nano_difference(struct timespec a, struct timespec b) {
+    return a.tv_nsec - b.tv_nsec + (a.tv_sec - b.tv_sec) * ONE_NANOSECOND;
 }
 
-int process_arguments(int ac, char **av, struct argument_parameters *arguments) {
-	return 0;
+struct timespec nano_add(struct timespec time, long nanos) {
+    struct timespec new_time = {};
+    new_time.tv_sec = time.tv_sec + (nanos / ONE_NANOSECOND);
+    new_time.tv_nsec = time.tv_nsec + (nanos % ONE_NANOSECOND);
+    if (new_time.tv_nsec >= ONE_NANOSECOND) {
+        new_time.tv_nsec -= ONE_NANOSECOND;
+        new_time.tv_sec++;
+    }
+    return new_time;
 }
 
 void calculate_framerate(struct frame_time_info *frames_data) {
-	frames_data->num_updates++;
-	struct timespec right_now = {};
-	clock_gettime(CLOCK_MONOTONIC, &right_now);
-
-	if (right_now.tv_sec - frames_data->prev_time.tv_sec >= 1) {
-		clock_gettime(CLOCK_MONOTONIC, &frames_data->prev_time);
-		frames_data->current_framerate = frames_data->num_updates;
-		frames_data->num_updates = 0;
-	}
+    frames_data->num_updates++;
+    struct timespec right_now = {};
+    clock_gettime(CLOCK_MONOTONIC, &right_now);
+
+    if (nano_difference(right_now, frames_data->prev_time) >= 1000000000) {
+        clock_gettime(CLOCK_MONOTONIC, &frames_data->prev_time);
+        frames_data->current_framerate = frames_data->num_updates;
+        frames_data->num_updates = 0;
+    }
 }
 
 // todo: change all timekeeping to nanoseconds.
 void frame_rate_cap(struct frame_time_info *frames_data) {
-	struct timespec wait = {};
-	struct timespec right_now = {};
-	clock_gettime(CLOCK_MONOTONIC, &right_now);
-	printf("beg\t %ld nanosec\n", frames_data->beginning_of_framrate_calc.tv_nsec);
-	printf("now\t %ld nanosec\n", right_now.tv_nsec);
-	printf("ns\t %ld \n", frames_data->ns_per_frame);
-	if (right_now.tv_sec - frames_data->beginning_of_framrate_calc.tv_sec == 1) {
-		wait.tv_nsec = frames_data->ns_per_frame - (right_now.tv_nsec - frames_data->beginning_of_framrate_calc.tv_nsec + 999999999);
-	} else {
-		wait.tv_nsec = frames_data->ns_per_frame - (right_now.tv_nsec - frames_data->beginning_of_framrate_calc.tv_nsec);
-	}
-
-	if (wait.tv_nsec < 0) {	 // already behind time, dont wait before the next update
-		wait.tv_nsec = 1;
-		// return;
-	}
-	nanosleep(&wait, NULL);
-
-	clock_gettime(CLOCK_MONOTONIC, &frames_data->beginning_of_framrate_calc);
-
-	//	printf("Wait:     %ld nano\n", wait.tv_nsec);
+    struct timespec wait = {};
+    struct timespec right_now = {};
+
+    struct timespec expected_frame_end_time =
+        nano_add(frames_data->prev_time,
+                 frames_data->num_updates * frames_data->ns_per_frame);
```

Comment:
The math is unclear here

Reply:
Appears to be a strange result whenever prev_time is updated as well. Time for a frame is inconsistent whenever prev_time is updated